src/frontend/src/optimizer/rule/top_n_to_vector_search_rule.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright 2025 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::assert_matches::assert_matches;
+
+use risingwave_common::types::DataType;
+use risingwave_common::util::sort_util::ColumnOrder;
+use risingwave_pb::common::PbDistanceType;
+
+use crate::expr::{Expr, ExprImpl, ExprRewriter, ExprType, InputRef};
+use crate::optimizer::plan_node::generic::TopNLimit;
+use crate::optimizer::plan_node::{
+    LogicalPlanRef as PlanRef, LogicalProject, LogicalVectorSearch, PlanTreeNodeUnary,
+};
+use crate::optimizer::rule::prelude::*;
+use crate::optimizer::rule::{BoxedRule, Rule};
+
+pub struct TopNToVectorSearchRule;
+
+impl TopNToVectorSearchRule {
+    pub fn create() -> BoxedRule<Logical> {
+        Box::new(TopNToVectorSearchRule)
+    }
+}
+
+impl Rule<Logical> for TopNToVectorSearchRule {
```

Comment:
I think you can draw a graph to represent how the plan is transformed before and after this rule. AI is good at this stuff.